File: notebookspython/08-triple_in_one.py
class TripleInOne:

    def __init__(self, stackSize: int):
        self.stack_size = stackSize
        self.array = [0] * (stackSize * 3)
        self.sizes = [0, 0, 0]

    def push(self, stackNum: int, value: int) -> str:
        # 如果当前栈满了，拒绝压入，返回友好提示
        if self.sizes[stackNum] == self.stack_size:
            msg = f"❌ 【エラー】{stackNum}号スタックは満杯です！値 [{value}] を追加できません。"
            print(msg)
            return msg
        
        # 计算绝对索引并存入
        abs_index = stackNum * self.stack_size + self.sizes[stackNum]
        self.array[abs_index] = value
        self.sizes[stackNum] += 1
        
        msg = f"📥 【Push成功】{stackNum}号スタックに値 [{value}] をプッシュしました。"
        print(msg)
        return msg

    def pop(self, stackNum: int) -> str:
        # 如果栈已经是空的，返回你指定的提示
        if self.sizes[stackNum] == 0:
            msg = f"⚠️ 【警告】{stackNum}号スタックは既に空です！ポップできません。"
            print(msg)
            return msg
        
        # 计数器减 1 并取出值
        self.sizes[stackNum] -= 1
        abs_index = stackNum * self.stack_size + self.sizes[stackNum]
        value = self.array[abs_index]
        
        msg = f"📤 【Pop成功】{stackNum}号スタックから要素 [{value}] を取り出しました。"
        print(msg)
        return msg

    def peek(self, stackNum: int) -> str:
        if self.sizes[stackNum] == 0:
            msg = f"👀 【Peek】{stackNum}号スタックは空なので、覗き見できる要素はありません。"
            print(msg)
            return msg
        
        abs_index = stackNum * self.stack_size + (self.sizes[stackNum] - 1)
        value = self.array[abs_index]
        
        msg = f"👀 【Peek】{stackNum}号スタックの先頭にある要素は [{value}] です。"
        print(msg)
        return msg

    def isEmpty(self, stackNum: int) -> str:
        if self.sizes[stackNum] == 0:
            msg = f"❓ 【状態確認】{stackNum}号スタックは「空」です。"
        else:
            msg = f"❓ 【状態確認】{stackNum}号スタックには現在 [{self.sizes[stackNum]}] 個の要素が入っています。"
        print(msg)
        return msg

if __name__ == "__main__":
    print("--- 🌟 三合一栈 交互日志开始 🌟 ---\n")
    
    # 1. 创建一个大小为 1 的三合一栈
    # 对应：["TripleInOne"] -> [[1]]
    obj = TripleInOne(1) 
    
    # 2. 往 0 号栈压入 1 -> 对应：["push"] -> [[0, 1]]
    obj.push(0, 1)
    
    # 3. 试图往 0 号栈再压入 2（应该失败，因为大小只有 1） -> 对应：["push"] -> [[0, 2]]
    obj.push(0, 2)
    
    # 4. 从 0 号栈弹出元素 -> 对应：["pop"] -> [[0]]
    obj.pop(0)
    
    # 5. 再次尝试从 0 号栈弹出元素（应该警告已空） -> 对应：["pop"] -> [[0]]
    obj.pop(0)
    
    # 6. 第三次尝试从 0 号栈弹出元素 -> 对应：["pop"] -> [[0]]
    obj.pop(0)
    
    # 7. 检查 0 号栈此时是不是空的 -> 对应：["isEmpty"] -> [[0]]
    obj.isEmpty(0)
    
    print("\n--- 🌟 运行结束，全流程可视化 🌟 ---")

File: notebooksc++/08-triple_in_one.cpp
#include <iostream>
#include <array>
#include <string>
#include <cassert>

using namespace std;

class TripleInOne {
public:
    TripleInOne(int stackSize) : stackSize(stackSize), array(stackSize * 3), sizes{0, 0, 0} {}

    string push(int stackNum, int value) {
        if (sizes[stackNum] == stackSize) {
            return "❌ 【错误】stack " + to_string(stackNum) + " 已满，无法 push " + to_string(value);
        }
        int index = stackNum * stackSize + sizes[stackNum];
        array[index] = value;
        sizes[stackNum]++;
        return "📥 【Push成功】stack " + to_string(stackNum) + " push " + to_string(value);
    }

    string pop(int stackNum) {
        if (sizes[stackNum] == 0) {
            return "⚠️ 【警告】stack " + to_string(stackNum) + " 已空，无法 pop";
        }
        sizes[stackNum]--;
        int index = stackNum * stackSize + sizes[stackNum];
        int value = array[index];
        return "📤 【Pop成功】stack " + to_string(stackNum) + " pop " + to_string(value);
    }

    string peek(int stackNum) {
        if (sizes[stackNum] == 0) {
            return "👀 【Peek】stack " + to_string(stackNum) + " 为空";
        }
        int index = stackNum * stackSize + sizes[stackNum] - 1;
        return "👀 【Peek】stack " + to_string(stackNum) + " top " + to_string(array[index]);
    }

    string isEmpty(int stackNum) {
        if (sizes[stackNum] == 0) {
            return "❓ 【状态】stack " + to_string(stackNum) + " 为空";
        }
        return "❓ 【状态】stack " + to_string(stackNum) + " 大小 " + to_string(sizes[stackNum]);
    }

private:
    int stackSize;
    vector<int> array;
    array<int, 3> sizes;
};

int main() {
    TripleInOne obj(1);

    string result1 = obj.push(0, 1);
    assert(result1.find("Push") != string::npos);

    string result2 = obj.push(0, 2);
    assert(result2.find("已满") != string::npos);

    string result3 = obj.pop(0);
    assert(result3.find("Pop成功") != string::npos);

    string result4 = obj.pop(0);
    assert(result4.find("已空") != string::npos);

    string result5 = obj.pop(0);
    assert(result5.find("已空") != string::npos);

    string result6 = obj.isEmpty(0);
    assert(result6.find("为空") != string::npos);

    cout << "All test cases passed ✅" << endl;
    return 0;
}
